### DISCUSSION TITLE
When synchronizing file clipboard, what should I set the fotmatId and data for client -> server?

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#8843] [Q&A]
Q1. What should fotmatId set when requesting clipboard sync to client -> server?

When I checked the source, I can't determine whether I should use something like `CF_HDROP`, `CFSTR_FILEDESCRIPTORW`, `CFSTR_FILECONTENTS`, or `FileGroupDescriptorW` again.

Q2. What format should the data be sent?

`Server -> Client `requested the contents of the file after receiving the file metadata.

What data format should I send to `Client->Server`?

I would like to know if I should send the contents of the file, metadata, or the entire data that changed the file to binary format, and which formId should I use in this case.

The text data was successfully synchronized to the `client -> server` clipboard using the format `UTF8_STRING`.

**Comment by akallabeth** [comment]
sorry, but what are you asking?
the questions can not be answered as it is not clear what you try to `sync`

general rule:
* there are some formats with a fixed `formatId`, like `CF_UNICODETEXT`, but most use dynamic ones (e.g. the source providing the `format list` also provides the `formatId` you can request it with)
* there is no format `UTF8_STRING` in `RDP`, that is `X11` specific ;)

**Reply by kimcoldwater** [reply]
I mean, what should I do when I send the clipboard file data from the client to the server.

I am implementing a project using JNI by referring to the Android version.( but not use Android )

```
		else if (event->type == EVENT_TYPE_CLIPBOARD)
		{
			UINT32 size;
			UINT32 formatId;
			ANDROID_EVENT_CLIPBOARD* clipboard_event = (ANDROID_EVENT_CLIPBOARD*)event;
			formatId = ClipboardRegisterFormat(afc->clipboard, "UTF8_STRING");
			size = clipboard_event->data_length;

			if (size)
				ClipboardSetData(afc->clipboard, formatId, clipboard_event->data, size);
			else
				ClipboardEmpty(afc->clipboard);

			android_cliprdr_send_client_format_list(afc->cliprdr);
			android_event_free((ANDROID_EVENT*)clipboard_event);
		}
```
Part of the code in `android_event.c`.

Sent a text data request as above and successfully changed the clipboard for the remote session to text data.

When I use `ClipboardSetData`, I need to put `formatId` and `data`, what `formatId` and data should I put in when I am a file?

**Reply by pnowack** [reply]
Oh no, you are completely misunderstanding the whole clipboard handling. `wClipboard` and its APIs (`ClipboardSetData`, `ClipboardGetData`, `ClipboardEmpty`, etc.) is just for converting actual clipboard data, it has nothing to do with the actual clipboard handling in RDP.
The actual clipboard handling is described in the documentation `[MS-RDPECLIP]`. It also describes the handling for files.

The actual relevant CLIPRDR API methods are (excluding here some regarding the initialization sequence):
`TempDirectory`
`FormatList`
`FormatListResponse`
`LockClipboardData`
`UnlockClipboardData`
`FormatDataRequest`
`FormatDataResponse`
`FileContentsRequest`
`FileContentsResponse`

`wClipboard` can partly handle files though and that is if the local peer is the shared owner of a uri-list. `wClipboard` then iterates through that list to create `FILEDESCRIPTORW` for each file and folder, which then just need to be serialized.
But as said above, it is just a way to convert data, it has nothing to do with the actual clipboard handling in RDP itself.

**Reply by kimcoldwater** [reply]
oh.. When I realized that I didn't understand something well, I checked the document and debugged it, and I realized the problem.

I think I can start implementing it now. Thank you for your kind advice.
